keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1618 @@
+as unschedulable at the moment.
+
+#### Filter
+
+This checks whether the given node has access to those ResourceClaims which
+were already allocated.
+
+For unallocated ResourceClaims with delayed allocation, only those nodes are
+filtered out that are explicitly listed in UnsuitableNodes or that don't match
+the optional AvailableOnNodes node filter.
+
+There are several
+reasons why such a deny list is more suitable than an allow list:
+- Nodes for which no information is available must pass the filter phase to be
+  included in the list that will be passed to the post-filter and to get copied
+  into the PotentialNodes field there.
+- A node can already be chosen while there is no information yet and, if
+  allocation for that node actually works, the Pod can get scheduled sooner.
+- Some resource drivers might not have any unsuitable nodes, for example
+  because they modify containers and that works on all nodes at all
+  times. Forcing such drivers to set an allow list would cause unnecessary
+  work.
+
+In its state for the Pod the scheduler plugin must remember when it rejected a
+node because of UnsuitableNodes. That information will be used in Post-filter
+to deallocate resources.
+
+#### Post-filter
+
+This is called when no suitable node could be found. If the Pod depends on ResourceClaims with delayed
+allocation, then deallocating one or more of these ResourceClaims may make the
+Pod schedulable after allocating the resource elsewhere. Therefore each
+ResourceClaim with delayed allocation is checked whether all of the following
+conditions apply:
+- allocated
+- not currently in use
+- it was the reason why some node could not fit the Pod, as recorded earlier in
+  Filter
+
+One of the ResourceClaims satisfying these criteria is picked randomly and its
+Phase is set to Deallocate. This may make it possible to run the Pod
+elsewhere. If it still doesn't help, deallocation may continue with another
+ResourceClaim, if there is one. To prevent deallocating all resources merely
+because the scheduler happens to check quickly, the next deallocation will only
+requested when there is none currently running.
+
+At the moment, the scheduler has no information that might enable it to
+prioritize which resource to deallocate first. Future extensions of this KEP
+might attempt to improve this.
+
+#### Pre-score
+
+This is passed a list of nodes that have passed filtering by the resource
+plugin and the other plugins. The PotentialNodes field of unallocated
+ResourceClaims with delayed allocation gets updated now if the field doesn't
+match the current list already.
+
+#### Reserve
+
+A node has been chosen for the Pod.
+
+If necessary, the SelectedNode field of ResourceClaims with delayed allocation
+gets set here and the scheduling attempt gets stopped for now. It will be
+retried when the ResourceClaim status changes.
+
+The Pod gets added to the ReservedFor field of its ResourceClaims to ensure that
```

Comment:
Who does this? kube-scheduler or the driver?

Reply:
The scheduler. Updated.